source/agora/script/Engine.d

```diff
@@ -0,0 +1,1257 @@
+
+        if (lock.bytes.length != Hash.sizeof)
+            return "LockType.Redeem requires 64-byte script hash in the lock script";
+        const Hash script_hash = Hash(lock.bytes);
+
+        Script unlock_script;
+        if (auto error = validateScript(ScriptType.Unlock, unlock.bytes,
+            this.StackMaxItemSize, unlock_script))
+            return error;
+
+        Stack stack = Stack(this.StackMaxTotalSize, this.StackMaxItemSize);
+        if (auto error = this.executeScript(unlock_script, stack, tx))
+            return error;
+
+        if (stack.empty())
+            return "LockType.Redeem requires unlock script to push a redeem script to the stack";
+
+        const redeem_bytes = stack.pop();
+        if (hashFull(redeem_bytes) != script_hash)
+            return "LockType.Redeem unlock script pushed a redeem script "
+                 ~ "which does not match the redeem hash in the lock script";
+
+        Script redeem;
+        if (auto error = validateScript(ScriptType.Redeem, redeem_bytes,
+            this.StackMaxItemSize, redeem))
+            return error;
+
+        if (auto error = this.executeScript(redeem, stack, tx))
+            return error;
+
+        if (this.hasScriptFailed(stack))
+            return "Script failed";
+
+        return null;
+    }
+
+    /***************************************************************************
+
+        Execute the script with the given stack and the associated spending
+        transaction. This routine may be called for all types of scripts,
+        lock, unlock, and redeem scripts.
+
+        An empty script will not fail execution. It's up to the calling code
+        to differentiate when this is an allowed condition.
+
+        Params:
+            script = the script to execute
+            stack = the stack to use for the script. May be non-empty.
+            tx = the associated spending transaction
+
+        Returns:
+            null if there were no errors,
+            or a string explaining the reason execution failed
+
+    ***************************************************************************/
+
+    private string executeScript (in Script script, ref Stack stack,
+        in Transaction tx) nothrow @safe
+    {
+        // tracks executable condition of scopes for use with IF / ELSE / etc
+        ScopeCondition sc;
+        const(ubyte)[] bytes = script[];
+        while (!bytes.empty())
+        {
+            OP opcode;
+            if (!bytes.front.toOPCode(opcode))
+                assert(0, "Script should have been syntactically validated");
+            bytes.popFront();
+
+            if (opcode.isConditional())
+            {
+                if (auto error = this.handleConditional(opcode, stack, sc))
+                    return error;
+                continue;
+            }
+
+            // whether the current scope is executable
+            // (all preceeding outer conditionals were true)
+            if (!sc.isTrue())
```

Comment:
This is problematic. Because it's not skipping the stack push opcodes. So if an `ELSE` condition turns the scope to false, we won't pop the right amount of opcodes and the data payload will be interpreted as an opcode, and fail with an assert.